pkg/controllers/cassandra/actions/scalein.go

```diff
@@ -0,0 +1,149 @@
+		}
+
+		allDecommissioned := true
+
+		nPilotsToRemove := int(*ss.Spec.Replicas - *a.NodePool.Replicas)
+		for i := 1; i <= nPilotsToRemove; i++ {
+			p := pilots[len(pilots)-i].DeepCopy()
+			if p.Spec.Cassandra == nil {
+				p.Spec.Cassandra = &v1alpha1.CassandraPilotSpec{}
+			}
+
+			if !p.Spec.Cassandra.Decommissioned {
+				p.Spec.Cassandra.Decommissioned = true
+				_, err := s.NavigatorClientset.NavigatorV1alpha1().Pilots(p.Namespace).Update(p)
+				if err != nil {
+					return err
+				}
+
+				s.Recorder.Eventf(
+					p,
+					corev1.EventTypeNormal,
+					a.Name(),
+					"Marked cassandra pilot %s for decommission", p.Name,
+				)
+			}
+
+			if p.Status.Cassandra == nil {
+				p.Status.Cassandra = &v1alpha1.CassandraPilotStatus{}
+			}
+
+			if !p.Status.Cassandra.Decommissioned {
+				allDecommissioned = false
+			}
+		}
+
+		if allDecommissioned {
+			ss.Spec.Replicas = a.NodePool.Replicas
+			_, err = s.Clientset.AppsV1beta1().StatefulSets(ss.Namespace).Update(ss)
+			if err == nil {
+				s.Recorder.Eventf(
+					a.Cluster,
+					corev1.EventTypeNormal,
+					a.Name(),
+					"All cassandra nodes decommissioned, scaling cluster to size %d", a.NodePool.Replicas,
+				)
+			}
+		}
+	}
+	if *ss.Spec.Replicas < *a.NodePool.Replicas {
+		return fmt.Errorf(
+			"the NodePool.Replicas value (%d) "+
+				"is greater than the existing StatefulSet.Replicas value (%d)",
+			a.NodePool.Replicas, *ss.Spec.Replicas,
+		)
+	}
+	return nil
+}
+
+func pilotNameForStatefulSetReplica(setName string, replica int32) string {
+	return fmt.Sprintf("%s-%d", setName, replica)
+}
+
+func pilotsForStatefulSet(state *controllers.State, cluster *v1alpha1.CassandraCluster, nodePool *v1alpha1.CassandraClusterNodePool, statefulSet *apps.StatefulSet) ([]*v1alpha1.Pilot, error) {
+	replicasPtr := statefulSet.Spec.Replicas
+	if replicasPtr == nil {
+		return nil, fmt.Errorf("statefulset.spec.replicas cannot be nil")
+	}
+	replicas := *replicasPtr
+	// TODO: read the cluster name and node pool name from the statefulset
+	// metadata instead of the Scale structure so we can make this a package
+	// function. For now, this way is safest until we are sure these
+	// labels are going to remain stable
+	selector, err := util.SelectorForNodePool(cluster, nodePool.Name)
+	if err != nil {
+		return nil, err
+	}
+	pilots, err := state.PilotLister.Pilots(cluster.Namespace).List(selector)
+	if err != nil {
+		return nil, err
+	}
+	var errs []error
+	var output []*v1alpha1.Pilot
+Outer:
+	for i := int32(0); i < replicas; i++ {
+		pilotName := pilotNameForStatefulSetReplica(statefulSet.Name, i)
+		for _, p := range pilots {
+			if p.Name == pilotName {
+				output = append(output, p)
+				continue Outer
+			}
+		}
+		errs = append(errs, fmt.Errorf("pilot %q not found", pilotName))
+	}
+	if len(errs) > 0 {
+		return nil, &k8sErrors.StatusError{
+			ErrStatus: metav1.Status{
+				Message: utilerror.NewAggregate(errs).Error(),
+				Reason:  metav1.StatusReasonNotFound,
+			},
+		}
```

Comment:
Never seen `k8sErrors.StatusError` used before. What's the purpose? I looked to see if we check for this error elsewhere, but couldn't see any such checks.